### DISCUSSION TITLE
Installing libmesh fails due to missing tirpc

### DISCUSSION THREAD

**Discussion by aturner-epcc** [idaholab/moose#28924] [Q&A Getting Started] [answered]
### Check these boxes if you have followed the posting rules.

- [X] I have consulted the [Posting Guidelines](https://github.com/idaholab/moose/discussions/18270).
- [X] I have searched the [Discussions Forum](https://github.com/idaholab/moose/discussions) and [MOOSE Framework Troubleshooting](https://mooseframework.inl.gov/help/troubleshooting.html) and have not found what I was looking for
- [X] Q&A Getting Started is the most appropriate category for my question (trouble installing, beginner user, ...)

### Issue or question about MOOSE

I am trying to install MOOSE from source on a CPU-based HPE Cray EX system (UK National Supercomputing Service, [ARCHER2](https://www.archer2.ac.uk)) and running into problems at the step to install libmesh. Following instructions at:

https://mooseframework.inl.gov/getting_started/installation/hpc_install_moose.html

I use:

```
module load PrgEnv-gnu
module load cray-python
module load cmake

./update_and_rebuild_libmesh.sh CC=cc CXX=CC FC=ftn
```

(`cc`, 'CC` and `ftn` are the compiler wrappers within the Cray programming environement, equivalent to `mpif90` etc. on other clusters.)

I get the following error:

```
checking for built-in XDR support... no
checking for XDR support in /usr/include/tirpc... no
configure: error: *** XDR was not found, but --enable-xdr-required was specified.
```

Indeed, that path does not exist on the system:

```
auser@ln03:|~> ls /usr/include/tirpc
ls: cannot access '/usr/include/tirpc': No such file or directory
```

I found the source for `libtirpc` on the internet at: https://www.linuxfromscratch.org/blfs/view/svn/basicnet/libtirpc.html and installed it myself at: `/work/z19/z19/auser/software/MOOSE/libtirpc` using the same compilers I am using for libmessh (and MOOSE).

I then added the following options to libmesh `./configure`:

```
--with-xdr-include=/work/z19/z19/auser/software/MOOSE/libtirpc/include/tirpc
--with-xdr-libname=/work/z19/z19/auser/software/MOOSE/libtirpc/lib/libtirpc.so
```

retrying the configure gives:

```
checking for XDR support in /work/z19/z19/auser/software/MOOSE/libtirpc/include/tirpc... no
configure: error: *** XDR was not found, but --enable-xdr-required was specified.
```

Any idea how I get past this issue?

**Comment by GiudGiud** [comment]
Hello

Did you install `libtirpc-devel` ?
Usually the cluster administrators will do that one for you

**Reply by aturner-epcc** [reply]
Thanks for the quick reply, much appreciated. System adminstrators will not generally change the login node image or install dev packages for requirements for individual user software (as that way madness lies, you could end up having to install extra things all the time as root when your service has over 3000 users as ARCHER2 does). Starting expectation is that dependencies can be satified in user space. Is there no way of satisfying this dependency from a user space install?

I see I could also install it using Spack:

https://packages.spack.io/package.html?name=libtirpc

but I assume this would be the same as my manual install. Why can't libmesh use an install of libtirpc that has not been installed by root?

**Reply by GiudGiud** [reply]
You most likely can, I just never have. 
My question was mostly did you install the devel version of that library

**Reply by aturner-epcc** [reply]
Ah, sorry, misunderstood! My local libtirpc install was from source so includes all headers and library files:

```
auser@ln04:/work/z19/z19/auser/software/MOOSE> ls libtirpc
etc  include  lib  share
auser@ln04:/work/z19/z19/auser/software/MOOSE> ls libtirpc/lib/
libtirpc.la  libtirpc.so  libtirpc.so.3  libtirpc.so.3.0.0  pkgconfig
auser@ln04:/work/z19/z19/auser/software/MOOSE> ls libtirpc/include/tirpc/
netconfig.h  rpc  rpcsvc
```

**Reply by GiudGiud** [reply]
@roystgnr it seems xdr is still missing. Is there another library we should be installing?

**Reply by GiudGiud** [reply]
If you just want to get moving you can modify update_and_rebuild_libmesh to remove the need for xdr. This won't affect anything in moose iirc (since we switched to ascii checkpoint)

**Comment by roystgnr** [comment, marked as answer]
In the future posting `config.log` would be helpful so we can see exactly why something isn't being detected.

This time, though, it's unfortunately already clear.  From `libmesh_optional_packages.m4`, "Note that we currently do not provide a way to specify a library PATH, but that could potentially be added."  The `--with-xdr-libname=foorpc` lets you try to link to `libfoorpc.so` instead of `libtirpc.so` if you have a different third-party RPC library, but for some reason we seem to be still assuming that the library is in your libs path already.

You can try `--with-xdr-include=/work/z19/z19/auser/software/MOOSE/libtirpc/include/tirpc LIBS='-L /work/z19/z19/auser/software/MOOSE/libtirpc/lib/ -Wl,-rpath -Wl,/work/z19/z19/auser/software/MOOSE/libtirpc/lib/'` (assuming those flags work for your compiler and linker; looks like your OS is SUSE-based so probably?)

I'll see if I can fix it in libMesh but it'll be a few weeks until that gets downstream to MOOSE.  In the meantime I'd also agree that (if my `LIBS` hack doesn't work) just disabling binary XDR is a decent workaround.  IMHO .xda.gz has always been a strictly better option.

**Reply by aturner-epcc** [reply]
@roystgnr @GiudGiud Thanks both for the helpful suggestions. For the moment, I have gone with disablign XDR support. If it turns out that is critical for people then I may revisit this.

**Reply by GiudGiud** [reply]
Sounds good thanks for letting us know.

**Reply by roystgnr** [reply]
The configure option you need, in https://github.com/libMesh/libmesh/pull/3985, just went into libMesh.  We should have it available in MOOSE in a week or two.

**Comment by acino79** [comment]
I'm trying to install moose on the same system (archer2). I managed to install libmesh, adding the path of a local install of tircp  addiing the the following  flag to _configure_libmesh.sh_:
```
--with-xdr-include="<workdir>/local/libtirpc/include/tirpc/" LIBS='-L<workdir>/local/libtirpc/lib/ -Wl,-rpath -Wl,<workdir>/local/libtirpc/lib/'
```
everything seems alright, but when I compile moose I get the following error :
```
[...]
/usr/bin/grep: /opt/local/libtirpc/lib/libtirpc.la: No such file or directory
/usr/bin/sed: can't read /opt/local/libtirpc/lib/libtirpc.la: No such file or directory
libtool:   error: '/opt/local/libtirpc/lib/libtirpc.la' is not a valid libtool archive
[...]
```

it seems that moose is still pointing to the wrong xdr library.  do you have any recommendation?

**Reply by roystgnr** [reply]
It's really odd that it's looking specifically for a .la archive when we didn't ask it to.  I don't have a libtirpc.la on my system either; `-ltirpc` finds the .so file and is happy with that.

It's also odd that you're getting that error "when I compile moose" - did the libMesh compilation succeed, but the MOOSE compilation failed?  AFAIK MOOSE shouldn't even know libtirpc exists, it should just be picking it up by copying the flags that we used in libMesh.

If you're using a current libMesh version, I'd love to know if the new option gives you any different behavior, though I admit this is a long shot.  Try `--with-xdr-include="<workdir>/local/libtirpc/include/tirpc/" --with-xdr-libdir="<workdir>/local/libtirpc/lib/"` (where I assume in this case `<workdir>` should just be `/opt`?) and let me know if you get anything different?

I do think you're not losing anything these days if you just give up and disable XDR, though.  IMHO .xda.gz and .cpa.gz (gzipped ASCII formats) are just as efficient as .xdr and .cpr (XDR-based binary equivalents) and are easier to work with, so for new "XDR" and checkpoint outputs MOOSE now defaults to the former and you only need to support the latter if you have previously-written files you still want to read without converting them first.

**Reply by GiudGiud** [reply]
> It's really odd that it's looking specifically for a .la archive when we didn't ask it to. I don't have a libtirpc.la on my system either; -ltirpc finds the .so file and is happy with that.

this feels like a `git clean -Xfd` is needed in moose/ and maybe in libmesh/ too

**Reply by acino79** [reply]
ok, I've tested again starting the installation from clean and I still see the same error.  
 
> If you're using a current libMesh version, I'd love to know if the new option gives you any different behavior, though I admit this is a long shot. Try `--with-xdr-include="<workdir>/local/libtirpc/include/tirpc/" --with-xdr-libdir="<workdir>/local/libtirpc/lib/"` (where I assume in this case `<workdir>` should just be `/opt`?) and let me know if you get anything different?

_<workdir>_ = is just the path to my personal work directory so, on /opt

**Comment by aturner-epcc** [comment]
@acino79 I documented the MOOSE build process on ARCHER2 at:

https://github.com/hpc-uk/build-instructions/blob/main/apps/MOOSE/ARCHER2_GCC.md

Note, this avoids the problem by removing support for XDR from the build.

**Reply by acino79** [reply]
@aturner-epcc thanks. I've tested and as expected fails all the tests related to XDR.

Can I ask why the changes to the fortran flag? I've tried linking umat and it seems that the Abaqus modules is missing now.

**Reply by aturner-epcc** [reply]
Because you are compiling with a newer version of GCC that, by default, checks that argument types match in subroutine calls. MPI Fortran subroutines do not conform to type matching (unless you are using the new interface with `use mpi_f08`) so this flag is needed to avoid compilation errors when compiling with newer GCC versions and using `use mpi`/`include mpif.h` syntax.
